precompiles/randomness/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+
+// #[cfg(test)]
+// mod mock;
+#[cfg(test)]
+mod tests;
+
+#[generate_function_selector]
+#[derive(Debug, PartialEq)]
+pub enum Action {
+	RelayBlockNumber = "relayBlockNumber()",
+	RelayEpochIndex = "relayEpochIndex()",
+	RequestBabeRandomnessCurrentBlock =
+		"requestBabeRandomnessCurrentBlock(address,uint256,uint64,bytes32,uint64)",
+	RequestBabeRandomnessOneEpochAgo =
+		"requestBabeRandomnessOneEpochAgo(address,uint256,uint64,bytes32)",
+	RequestBabeRandomnessTwoEpochsAgo =
+		"requestBabeRandomnessTwoEpochsAgo(address,uint256,uint64,bytes32)",
+	RequestLocalRandomness = "requestLocalRandomness(address,uint256,uint64,bytes32,uint64)",
+	FulfillRequest = "fulfillRequest(uint64)",
+	IncreaseRequestFee = "increaseRequestFee(uint64,uint256)",
+	ExecuteRequestExpiration = "executeRequestExpiration(uint64)",
+}
+
+pub const FULFILLMENT_ESTIMATED_COST: u64 = 1000u64; // TODO: get real value from benchmarking
+pub const LOG_SUBCALL_SUCCEEDED: [u8; 32] = keccak256!("SubcallSucceeded");
+pub const LOG_SUBCALL_FAILED: [u8; 32] = keccak256!("SubcallFailed");
+
+pub fn log_subcall_succeeded(address: impl Into<H160>) -> Log {
+	log0(address, LOG_SUBCALL_SUCCEEDED)
+}
+
+pub fn log_subcall_failed(address: impl Into<H160>) -> Log {
+	log0(address, LOG_SUBCALL_FAILED)
+}
+
+/// Reverts if fees and gas_limit are not sufficient to make subcall and cleanup
+fn ensure_can_provide_randomness<Runtime>(
+	code_address: H160,
+	gas_limit: u64,
+	request_fee: BalanceOf<Runtime>,
+	clean_up_cost: u64,
+) -> EvmResult<()>
+where
+	Runtime: pallet_randomness::Config + pallet_evm::Config + pallet_base_fee::Config,
+	BalanceOf<Runtime>: Into<U256>,
+{
+	// assert fee > gasLimit * base_fee
+	let gas_limit_as_u256: U256 = gas_limit.into();
+	if let Some(gas_limit_times_base_fee) =
+		gas_limit_as_u256.checked_mul(pallet_base_fee::Pallet::<Runtime>::base_fee_per_gas())
+	{
+		if gas_limit_times_base_fee >= request_fee.into() {
+			return Err(revert(
+				"Gas limit at current price must be less than fees allotted",
+			));
+		}
+	} else {
+		return Err(revert("Gas limit times base fee overflowed U256"));
+	}
+	let log_cost = log_subcall_failed(code_address)
+		.compute_cost()
+		.map_err(|_| revert("failed to compute log cost"))?;
+	// Cost of the call itself that the batch precompile must pay.
+	let call_cost = call_cost(U256::zero(), <Runtime as pallet_evm::Config>::config());
+	// assert gasLimit > overhead cost
+	let overhead = call_cost + log_cost + clean_up_cost;
+	if gas_limit <= overhead {
+		return Err(revert("Gas limit must exceed overhead call cost"));
+	}
+	Ok(())
+}
+
+/// Subcall to provide randomness
+/// caller must call `ensure_can_provide_randomness` before calling this function
+fn provide_randomness(
+	handle: &mut impl PrecompileHandle,
+	gas_limit: u64,
+	contract: H160,
+	randomness: H256,
+) -> EvmResult<()> {
+	let (reason, _) = handle.call(
+		contract,
+		None,
+		EvmDataWriter::new().write(randomness).build(),
```

Comment:
Missing selector. It should call a function of the contract with the randomness as the parameter `acceptRandomness(bytes32)` ?